395 implement new deterministic metrics kge

samland1116 opened the pull request:
This PR updates the metrics framework for KGE, KGEmod1, and KGEmod2 to allow for user defined constants. This PR also explores the impact of wrapper functions on query performance. By creating a wrapper around the KGE functions that ingests the KGE base models, we can specify user defined constants like below:

```
from teehr import DeterministicMetrics

kge = DeterministicMetrics.KlingGuptaEfficiency()
kge.sr = 0.5
kge.sa = 0.5
kge.sb = 0.5

include_metrics = [kge]

metrics_df = ev.metrics.query(
    group_by=["primary_location_id", "configuration_name"],
    include_metrics=include_metrics
).to_pandas()
```

Alternatively, the existing method to query metrics still works despite the presence of a wrapper, shown below:

```
from teehr import DeterministicMetrics

metrics_df = ev.metrics.query(
    group_by=["primary_location_id", "configuration_name"],
    include_metrics=[DeterministicMetrics.KlingGuptaEfficiency()]
).to_pandas()
```

Unexpectedly, the presence of a wrapper that _does not alter values_ within a given function results in somewhat significant performance improvements while the presence of a wrapper that _does alter values_ within a given function results in performance similar to that of our existing metrics implementation. 

I propose we implement this functionality on all existing metrics functions to standardize metrics behavior while simultaneously improving metric query performance.

samlamont reviewed (commented):
I think it looks go to me overall.  I wonder if we should just implement the wrappers for all metrics in this PR since it would get rid of the `requires_wrapper` checks and allow us to call the pandas_udf in the same way for all metrics?

samland1116 commented:
Adds the wrapper method for the following...

- Deterministic Metrics: 
  - mean_error
  - relative_bias
  - mean_absolute_relative_error
  - multiplicative_bias
  - pearson_correlation
  - r_squared
  - max_value_delta
  - annual_peak_relative_bias
  - spearman_correlation
  - _mean_error
    - received no update as this is a helper function. Moved to the top of the file for clarity.
  - KGE
  - KGE_mod1
  - KGE_mod2 
  - NSE
  - NSE_normalized
  - mean_absolute_error
  - root_mean_squared_error
    -  This method was utilized in 'root_mean_standard_deviation_ratio' so I created an identical helper function: '_root_mean_squared_error'. Moved to the top of the file for clarity.
  - root_mean_standard_deviation_ratio
  - max_value_timedelta

- Signature Metrics:
  - max_value_time
  - variance
  - count
  - minimum
  - maximum
  - average
  - sum  

Also removes the 'requires_wrapper' attribute from any impacted Metric Basemodels. Removes 'requires_wrapper' conditional statements from aggregation functions/bootstrapping functions.

samlamont reviewed (approved):
Looks good to me, tests pass for me.  Just double check, no updates are necessary to the documentation since these changes happen in the background correct?
